ksonnet-gen/jsonnet/rewrite.go

```diff
@@ -0,0 +1,128 @@
+// for `{error: ...}`, the `FieldKey` would be `"error"` (with
+// quotation marks, to avoid collisions).
+type FieldKey string
+
+// FuncParam represents the parameter to a Jsonnet function, after
+// being rewritten to avoid collisions with Jsonnet keywords and
+// normalized to fit the Jsonnet style (i.e., lowerCamelCase) using a
+// manual set of custom transformations that change per Kubernetes
+// version. For example, in `foo(BarAPI) {...}`, `FuncParam` would be
+// `barApi`, and in `foo(error) {...}`, `FuncParam` would be
+// `errorParam`.
+type FuncParam string
+
+// Identifier represents any identifier in a Jsonnet program, after
+// being normalized to fit the Jsonnet style (i.e., lowerCamelCase)
+// using a manual set of custom transformations that change per
+// Kubernetes version. For example, `fooAPI` becomes `fooApi`.
+type Identifier string
+
+// RewriteAsFieldKey takes a `PropertyName` and converts it to a valid
+// Jsonnet field name. For example, if the `PropertyName` has a value
+// of `"error"`, then this would generate an invalid object, `{error:
+// ...}`. Hence, this function will quote this string, so that it ends
+// up like: `{"error": ...}`.
+func RewriteAsFieldKey(text kubespec.PropertyName) FieldKey {
+	// NOTE: Because the field needs to have precisely the same text as
+	// the Kubernetes API spec, we do not compute a version-specific ID
+	// alias as we do for other rewrites.
+	if _, ok := jsonnetKeywordSet[text]; ok {
+		return FieldKey(fmt.Sprintf("\"%s\"", text))
+	}
+	return FieldKey(text)
+}
+
+// RewriteAsFuncParam takes a `PropertyName` and converts it to a
+// valid Jsonnet function parameter. For example, if the
+// `PropertyName` has a value of `"error"`, then this would generate
+// an invalid function parameter, `function(error) ...`. Hence, this
+// function will alter the identifier, so that it ends up like:
+// `function(errorParam) ...`.
+//
+// NOTE: This transformation involves a hand-curated style change to
+// lowerCamelCase (e.g., `fooAPI` -> `fooApi`). This list changes per
+// Kubernetes version, according to identifiers that don't conform to
+// this style.
+func RewriteAsFuncParam(
+	k8sVersion string, text kubespec.PropertyName,
+) FuncParam {
+	id := RewriteAsIdentifier(k8sVersion, string(text))
+	if _, ok := jsonnetKeywordSet[kubespec.PropertyName(id)]; ok {
+		return FuncParam(fmt.Sprintf("%sParam", id))
+	}
+	return FuncParam(id)
+}
+
+// RewriteAsIdentifier takes a `GroupName`, `ObjectKind`,
+// `PropertyName`, or `string`, and converts it to a Jsonnet-style
+// Identifier. Typically this includes lower-casing the first letter,
+// but also changing initialisms like fooAPI -> fooApi.
+//
+// NOTE: This transformation involves a hand-curated style change to
+// lowerCamelCase (e.g., `fooAPI` -> `fooApi`). This list changes per
+// Kubernetes version, according to identifiers that don't conform to
+// this style.
+func RewriteAsIdentifier(
```

Comment:
Wasn't sure what the right way to do this is, because I don't often see people passing strings around, and to call `Stringer#String` in that method I think you have to turn it into a pointer. So we have a lot of methods using the pointer to value as the receiver. If that's not right, we can fix it.

```go
func (ok ObjectKind) String() string {
	return string(ok)
}
```